crates/sui-core/src/execution_scheduler/execution_scheduler_impl.rs

```diff
@@ -630,6 +656,110 @@ impl ExecutionScheduler {
+            // Fastpath transactions that perform object balance withdraws
+            // must wait for consensus to assign the accumulator version.
+            // We cannot optimize the scheduling by processing fastpath object withdraws
+            // sooner because these may get reverted, and we don't want them
+            // pollute the scheduler tracking state.
+            // TODO: We could however optimize execution by caching
+            // the execution state to avoid re-execution.
+            return false;
+        };
+        match scheduler.schedule(object_withdraws, accumulator_version) {
+            // Sufficient balance, we can go ahead and commit the execution results as it is.
+            ObjectBalanceWithdrawStatus::SufficientBalance => true,
+            // Currently insufficient balance. We need to wait until it reach a deterministic state
+            // before we can determine if it is sufficient.
+            // At that time we will have to re-enqueue the transaction for execution again.
+            // Re-enqueue is handled here so the caller does not need to worry about it.
+            ObjectBalanceWithdrawStatus::Pending(receiver) => {
+                let scheduler = self.clone();
+                let cert = certificate.clone();
+                let execution_env = execution_env.clone();
+                tokio::task::spawn(async move {
+                    // TODO: Is this unwrap safe?
+                    let result = receiver.await.unwrap();
+                    match result {
+                        BalanceWithdrawStatus::Unknown => {
+                            // The withdraw is now deterministically sufficient,
+                            // so we can enqueue the transaction again and it will be able
+                            // to go through in the next execution.
+                            // TODO: We should be able to optimize this by avoiding re-execution.
+                            scheduler.send_transaction_for_execution(
+                                &cert,
+                                execution_env,
+                                // TODO: Fix enqueue_time.
+                                Instant::now(),
+                            );
+                        }
+                        BalanceWithdrawStatus::InsufficientBalance => {
+                            // Re-enqueue with insufficient balance status, so it will be executed
+                            // in the next execution and fail through early error.
+                            let execution_env = execution_env.with_insufficient_balance();
+                            scheduler.send_transaction_for_execution(
+                                &cert,
+                                execution_env,
+                                // TODO: Fix enqueue_time.
+                                Instant::now(),
+                            );
```

Comment:
looks like you could factor out the `send_transaction_for_execution` call and only use the match stmt to construct `execution_env`